Add refreshed nuget staging proposal (#14978)
NuGet/Home: accepted/2026/package-staging.md

Staging the first version of a new package ID reserves the ID permanently, same as a normal push. If that staged version is later deleted or expires, the ID remains reserved.

The Upload page also supports staging via a Publishing Mode option:

![Upload Package - Publishing Mode](images/upload-package.png)
*Figure 1. Upload Package page with Publishing Mode selector.*

#### Staging groups

Packages can optionally be organized into staging groups. A group is a named collection of staged packages owned by a user or organization.

```
dotnet nuget stage group create my-release --name "My Release"
dotnet nuget stage push MyPackage.1.0.0.nupkg --group my-release
```

##### Managing groups

Groups are useful for coordinating releases across many packages. A team can stage hundreds of packages into a single group over several days, verify all of them pass validation, and then promote the entire group at once.

> Groups are ephemeral. They are deleted after a successful promotion and expire after a configurable TTL. Each release gets its own group.

Author: The current design is 30 days

# Package Staging

- Author: James Parsons ([@japarson](https://github.com/japarson))
- GitHub Issue: [NuGet/NuGetGallery#3931](https://github.com/NuGet/NuGetGallery/issues/3931)

## Summary

> Package staging lets authors push packages to nuget.org ahead of time, run them through full validation (malware scanning, signing), and promote them later with a single action through the Gallery UI. Staged packages are invisible to consumers until promoted. Promotion requires that the user is logged into nuget.org with two-factor authentication enabled on their account.

Reviewer: ```suggestion
Package staging lets authors push packages to nuget.org ahead of time, run them through full validation (malware scanning, signing), and promote them later with a single action through the Gallery UI. Staged packages are invisible to consumers until promoted. Promotion is only available for users with two-factor authentication enabled.
```

This builds on the [2024 Release Staging and Deprecation proposal](../2024/release-staging-and-deprecation.md) with a narrower scope focused on staging and promotion, informed by direct feedback from the .NET release team and security requirements around supply chain protection.

## Motivation

### Supply chain protection

> If a CI/CD pipeline is compromised or credentials leak, an attacker can push malicious packages to nuget.org today. Staging introduces a separation between pushing and promoting. CI/CD can only stage packages. A human must log into nuget.org with 2FA enabled to promote them. Even if an attacker compromises a build pipeline, they can only stage invisible packages that never reach consumers.

Reviewer: This claim is unclear/unsubstantiated. Is the suggestion that you can switch your account to staging only? The text doesn't say that.

Author: It's mentioned later that there will be a new staging-only API scope, but I'll add it here as well.

> Staging is entirely opt-in. The existing `dotnet nuget push` command and push API continue to work exactly as they do today, including for older SDK versions. Staging uses a separate endpoint that newer SDKs may expose through commands such as `dotnet nuget stage push`. The exact command names and syntax shown in this proposal are illustrative and not final.

Reviewer: ```suggestion
Staging is opt-in. The existing `dotnet nuget push` command and push API continue to work exactly as they do today, including for older SDK versions. Staging uses a separate endpoint that newer SDKs may expose through commands such as `dotnet nuget stage push`. The exact command names and syntax shown in this proposal are illustrative and not final.
```

### Security embargo timing

.NET security releases are coordinated with public disclosure of vulnerabilities. The team cannot push packages before 10 AM Pacific because early availability would give attackers time to analyze changes and derive exploits before users can patch. But validation takes 40 to 60 minutes after push, so the packages are not available until well after disclosure. This creates a window where the vulnerability is publicly known but the fix is not yet restorable.

> Staging eliminates this gap. Packages are pre-validated and sitting in a staged state. At the coordinated disclosure time, promotion makes them available without waiting for validation.

Reviewer: ```suggestion
Staging eliminates this gap. Packages are pre-validated, in a staged state. At the coordinated disclosure time, promotion makes them available without waiting for validation.
```

### Pre-validation

> The .NET team pushes 2000+ packages to nuget.org every Patch Tuesday. Validation (malware scanning, certificate checks, repository co-signing) takes 40 to 60 minutes for that volume. If a package fails validation, the team discovers it on release day and has to rebuild, re-sign, and re-push during a tight embargo window.

Reviewer: ```suggestion
The .NET team pushes 2000+ packages to nuget.org every Patch Tuesday. Validation (malware scanning, certificate checks, repository co-signing) takes 40 to 60 minutes for that volume. If a package fails validation, the team discovers it on release day and has to rebuild, re-sign, and re-push, delaying the release of the patches but not the disclosure. Delay can create a 0-day opportunity.
```

Reviewer: The main point is that embargo is blown. MSRC may publish disclosure before we do.

When a user clicks "Promote" on a staged package or group, the user must be logged into nuget.org with two-factor authentication enabled on their account.

For groups, the owner promotes the entire group at once. If any package in the group is still validating or failed validation, the group cannot be promoted until the issue is resolved.

Promotion runs asynchronously, and owners can refresh the page to view progress. Group members are processed independently, so a group may partially succeed; failed items remain visible for correction.

The Gallery keeps an owner-visible history of package and group promotions, including their outcomes.

After promotion, the packages enter the normal V3 pipeline and become restorable through the standard process. The public "published" timestamp is set at promotion time, not at stage time.

The package detail page shows staged packages to the owner with validation status, promote, download (for inspection), replace, and delete actions.

![Package Display - Staged Package](images/display-package.png)
*Figure 4. Package detail page for a staged package showing promotion and staging group controls.*

#### Trusted Publishing integration

> [Trusted Publishing](../2024/trusted-publishers-oidc-for-nuget-push.md) (OIDC) is the recommended authentication method for staging, but API keys are also supported. A new "Stage packages" API key scope is added alongside Push, Push only new versions, and Unlist.

Reviewer: Is there a strong reason to support API keys if we're trying to get rid of them? Or is it more work to dissallow them?

Author: We're still in the beginning phase of phasing out api keys and there are still many users not onboarded to trusted publishing who can benefit from staging. For example, Trusted Publishing support for GitLab pipelines was only just recently (?) released.

![API Key Scopes - Stage packages](images/api-key-scopes.png)

*Figure 5. API key management showing the new "Stage packages" scope.*

> Trusted Publishing configurations support a "Restrict to staging only" scope. When configured with this scope, a GitHub Actions workflow (or other supported CI/CD provider) can stage packages but cannot push directly to the public feed. This is enforced in both directions: the staging endpoints accept staging-only credentials, and the normal push endpoint rejects them with 403.

Reviewer: > Trusted Publishing configurations support a "Restrict to staging only" scope. 

Nice!

> Trusted Publishing configurations support a "Restrict to staging only" scope. When configured with this scope, a GitHub Actions workflow (or other supported CI/CD provider) can stage packages but cannot push directly to the public feed. This is enforced in both directions: the staging endpoints accept staging-only credentials, and the normal push endpoint rejects them with 403.

Reviewer 2: How is it different from selecting only "Stage packages" option in API key scope settings? Why do we need another option here?

Author: I edited the doc to more clearly delineate that api keys will have staging scope and trusted publishing will have a staging only "constraint".

# Package Staging

- Author: James Parsons ([@japarson](https://github.com/japarson))
- GitHub Issue: [NuGet/NuGetGallery#3931](https://github.com/NuGet/NuGetGallery/issues/3931)

## Summary

> Package staging lets authors push packages to nuget.org ahead of time, run them through full validation (malware scanning, signing), and publish them later with a single action through the Gallery UI. Staged packages are invisible to consumers until published. Publishing requires logging into nuget.org, which enforces two-factor authentication as a proof of presence check.

Reviewer 3: nit: it would be nice if we could have a sentence per line. Makes commenting during reviews much easier.

This builds on the [2024 Release Staging and Deprecation proposal](../2024/release-staging-and-deprecation.md) with a narrower scope focused on staging and publishing, informed by direct feedback from the .NET release team and security requirements around supply chain protection.

## Motivation

### Supply chain protection

> If a CI/CD pipeline is compromised or credentials leak, an attacker can push malicious packages to nuget.org today. Staging introduces a separation between pushing and publishing. CI/CD can only stage packages. A human must log into nuget.org (which requires two-factor authentication) to publish them. Even if an attacker compromises a build pipeline, they can only stage invisible packages that never reach consumers.

Reviewer 3: what happens if a customer runs `dotnet nuget push`(How does the server react to these requess)? They could be on an old dotnet sdk.

.NET security releases are coordinated with public disclosure of vulnerabilities. The team cannot push packages before 10 AM Pacific because early availability would give attackers time to analyze changes and derive exploits before users can patch. But validation takes 40 to 60 minutes after push, so the packages are not available until well after disclosure. This creates a window where the vulnerability is publicly known but the fix is not yet restorable.

Staging eliminates this gap. Packages are pre-validated and sitting in a staged state. At the coordinated disclosure time, publishing makes them available without waiting for validation.

### Pre-validation

The .NET team pushes 2000+ packages to nuget.org every Patch Tuesday. Validation (malware scanning, certificate checks, repository co-signing) takes 40 to 60 minutes for that volume. If a package fails validation, the team discovers it on release day and has to rebuild, re-sign, and re-push during a tight embargo window.

Staging lets authors push packages days or weeks early. Validation runs during that window. Failures are caught and fixed before release day, not during it.

### Publish latency

> Today, the time from push to all packages being restorable includes validation plus the V3 pipeline (catalog, flat container, registration, search). For 2000 packages, total time was about 2 hours during the March 2026 Patch Tuesday. By removing validation from the critical path, the only remaining cost is the V3 pipeline, which should bring the total time from publish to restorable down significantly.

Reviewer 4: We should call out that we made work to bring the V3 ingestion down from 1 hour to around 5 minutes.  So I overall, we will be able to manage to bring down the published time from 2 hours to only 5 min, assuming package are properly staged and validated prior to release

## Explanation

#### Staging a package

```mermaid
flowchart LR
    Push["Stage Push"] --> Validation
    Validation -->|success| Staged
    Validation -->|failure| FailedValidation
    Staged -->|publish via Gallery| Available
    Available --> V3["V3 Pipeline"]
```

Authors push packages to nuget.org using a new staging endpoint. The CLI command looks like:

```
dotnet nuget stage push MyPackage.1.0.0.nupkg
```

> The package goes through the same validation pipeline as a normal push: malware scanning, author signature checks, certificate revocation checks, and repository co-signing. The difference is what happens after validation succeeds. Instead of becoming publicly available, the package enters a "staged" state. It is not restorable, not searchable, and not visible on nuget.org to anyone except the package owner.

Reviewer 4: I don't think the package are visible at all even for the package owners.  Package Owner would only be able to see the staged package from the UI,   not through search. Let's clarify that.

If validation fails, the author is notified the same way as a normal push failure. They can fix the issue and re-push. Since this happens days or weeks before release, there is no time pressure.

Re-pushing the same content is a no-op. Pushing different content for the same ID and version replaces the staged package and re-runs validation.

#### Staging groups

Packages can optionally be organized into staging groups. A group is a named collection of staged packages owned by a user or organization.

```

> dotnet nuget stage group create my-release --name "My Release"

Reviewer 4: I don't think we need to have a create command.  Let's simplify this!

If you do 'dotnet nuget stage --group "BCL Release 11.0" , the push will automatically stage the package under that group.  If you do not supply the --group,  the package will be stage under no group

For individual packages, the owner navigates to their package management page, finds the staged package, and clicks "Publish." For groups, the owner publishes the entire group at once. If any package in the group is still validating or failed validation, the group cannot be published until the issue is resolved.

After publishing, the packages enter the normal V3 pipeline and become restorable through the standard process.

#### Trusted Publishing integration

[Trusted Publishing](../2024/trusted-publishers-oidc-for-nuget-push.md) (OIDC) is the recommended authentication method for automated package publishing on nuget.org. API keys are strongly discouraged and should be replaced with Trusted Publishing.

Trusted Publishing configurations will support a new "stage only" permission scope. When configured with this scope, a GitHub Actions workflow (or other supported CI/CD provider) can stage packages but cannot publish them. This aligns with the security model: CI/CD stages, humans publish.

#### Gallery UI

The nuget.org Manage Packages page adds two new sections:

> ![Manage Packages - Staged Packages and Staging Groups](images/staging-manage-packages.png)

Reviewer 4: I think we should only have one category: "Staged Packages"  When expanding Stage Packages, staged packages would be grouped by "Groups", one of the group being "none".  It's important for the groups to be expanded  in order to see the packages visible under that group.  For instance, if a package fails to validate, the ability to re-kick a validation needs to be available, even for packages under a group

| V3 flat container / registration / catalog | No |
| API package metadata | No |

Staged packages that are never published are automatically cleaned up after 30 days. This is longer than npm's 72-hour expiration because some release cycles are monthly, and teams may stage packages 2 to 3 weeks before release.

For grouped packages, the TTL is on the group, not individual packages. It resets whenever a new package is pushed to the group. When the TTL expires, the entire group and all its packages are deleted. Owners receive email notifications 7 days before expiration and when packages are deleted.

Groups are deleted after a successful publish since they have served their purpose. Empty groups persist until TTL or manual deletion.

#### Limits

- Max 20 staging groups per owner
- Max 1000 packages per group
- Max 5000 ungrouped staged packages per owner

#### CLI commands

All commands discover the staging endpoint via the `PackageStaging/1.0.0` resource type in the service index. Sources that do not advertise this resource return an error indicating staging is not supported.

All commands support `--format json` for machine-readable output.

> **`dotnet nuget stage push`**

Reviewer 3: Question: I see we are adding a new command. What happens to the exisitng `dotnet nuget push` command?

I would assume we won't be deprecating it as it would still be used by other package sources.

Reviewer 4: we do not change the behavior for 'dotnet nuget push'

Push a package to staging. The package goes through validation and transitions to Staged on success.

```
dotnet nuget stage push <PACKAGE_PATH> [options]

Arguments:
  PACKAGE_PATH    Path to the .nupkg file to stage.

Options:
  -s, --source <SOURCE>     The package source to use. Defaults to nuget.org.
  -k, --api-key <KEY>       The API key for the source. Not recommended; use Trusted Publishing instead.

>   --group <GROUP_ID>        Add the package to a staging group. The group must already exist.

Reviewer 4: I think the requirements of creating the group first is an overhead.  if you pass --group "Group_ID", if the group doesn't already exist, it will create it

  --format <FORMAT>         Output format: text (default) or json.
```

Examples:
```
dotnet nuget stage push MyPackage.1.0.0.nupkg
dotnet nuget stage push MyPackage.1.0.0.nupkg --group my-release
```

**`dotnet nuget stage list`**

List staged packages for the authenticated user.

```

> dotnet nuget stage list [options]

Reviewer 4: to be clear, 'dotnet nuget stage list' will list all stage package by group

Options:
  -s, --source <SOURCE>     The package source to use. Defaults to nuget.org.
  -k, --api-key <KEY>       The API key for the source.
  --group <GROUP_ID>        Filter to packages in a specific group.
  --format <FORMAT>         Output format: text (default) or json.
```

Example output:
```
Staged Packages (3 packages)

Package ID                                Version              Group            Status     Expires
Microsoft.Extensions.Logging              10.0.0-preview.4     my-release       Ready      Jul 10, 2026
Microsoft.Extensions.DependencyInjection  10.0.0-preview.4     my-release       Ready      Jul 10, 2026
Contoso.Utilities                         3.1.0                (ungrouped)      Validating Jul 8, 2026
```

**`dotnet nuget stage delete`**

Delete a staged package. Fails if the package is currently validating.

```

> dotnet nuget stage delete <PACKAGE_ID> <VERSION> [options]

Reviewer 4: I which context this command would be needed?

Arguments:
  PACKAGE_ID      The package ID to delete.
  VERSION         The package version to delete.

Options:
  -s, --source <SOURCE>     The package source to use. Defaults to nuget.org.
  -k, --api-key <KEY>       The API key for the source.
  --no-confirm              Skip the confirmation prompt.
  --format <FORMAT>         Output format: text (default) or json.
```

**`dotnet nuget stage group create`**

> Create a new staging group.

Reviewer 4: I don't think we need that command.

```
dotnet nuget stage group create <GROUP_ID> [options]

Arguments:
  GROUP_ID        The group identifier. Lowercase letters, numbers, and dashes only.

Options:
  -s, --source <SOURCE>     The package source to use. Defaults to nuget.org.
  -k, --api-key <KEY>       The API key for the source.
  --name <DISPLAY_NAME>     Optional display name for the group. Defaults to the group ID.
  --format <FORMAT>         Output format: text (default) or json.
```

**`dotnet nuget stage group list`**

List staging groups for the authenticated user.

```
dotnet nuget stage group list [options]

Options:
  -s, --source <SOURCE>     The package source to use. Defaults to nuget.org.
  -k, --api-key <KEY>       The API key for the source.
  --format <FORMAT>         Output format: text (default) or json.
```

Example output:
```
Staging Groups (2 groups)

> Group ID           Name                Packages   Status      Expires

Reviewer 4: (Ungroup) should an implied group for ungrouped packages

my-release         My Release          2 ready    Ready       Jul 10, 2026
hotfix-june        June Hotfix         0          Empty       Jul 1, 2026
```

**`dotnet nuget stage group delete`**

> Delete a staging group and all its staged packages. Fails if any package in the group is currently validating.

Reviewer 4: Maybe a --force command to force deletion of a group no matter what?

```
dotnet nuget stage group delete <GROUP_ID> [options]

Arguments:
  GROUP_ID        The group identifier to delete.

Options:
  -s, --source <SOURCE>     The package source to use. Defaults to nuget.org.
  -k, --api-key <KEY>       The API key for the source.
  --no-confirm              Skip the confirmation prompt.
  --format <FORMAT>         Output format: text (default) or json.
```

## Drawbacks

- **Publish is not automatable.** Requiring Gallery UI login means publishing cannot be scripted or run from CI/CD. This is intentional for security but adds a manual step to release workflows. Someone has to log in and click "Publish" on release day.

> - **No atomic publish.** We are not guaranteeing that all packages in a group become restorable at the exact same instant. The V3 pipeline processes packages in batches and there will be a small window (minutes) where some packages are available and others are not. Package authors who need strict ordering should manage dependency sequencing on their side.

Reviewer 4: By dependencies sequencing, this means proper grouping.   For instance, BCLs groups should be release first, then SDK Libraries and then Tools

Re-pushing the same content is a no-op (safe for CI/CD retries). Pushing different content for the same ID and version replaces the staged package, re-runs validation, and resets the expiration timer. Replacing a grouped package keeps its group membership.

Symbol packages (`.snupkg`) can also be staged. The parent `.nupkg` must already be staged or validating. Symbol validation and ingestion are deferred until promotion.

Staging the first version of a new package ID reserves the ID permanently, same as a normal push. If that staged version is later deleted or expires, the ID remains reserved.

The Upload page also supports staging via a Publishing Mode option:

![Upload Package - Publishing Mode](images/upload-package.png)
*Figure 1. Upload Package page with Publishing Mode selector.*

#### Revalidation

Validation results go stale over time. Malware signature databases update daily and certificate revocation status can change. A package that passed validation weeks ago may no longer be clean.

> After a configurable staleness window (default: a few days), the package transitions to a "Stale" state in the UI. Stale packages cannot be promoted until the owner triggers revalidation. Only external-state checks rerun: certificate revocation and malware scanning. Content-only checks (signature structure, package format) are skipped since the bytes have not changed.

Reviewer 5: Have you considered revalidating automatically? 
At least like once or twice. 
This would allow a Tuesday release to be able to be prepped on like Thursday/Friday.

Maybe like revalidate if that package had been pushed within a week?

I get that it creates a bit of a potential for race conditions, but just an idea.

Author: I updated the spec to clarify the staleness window is 7 days which should avoid weekend scenarios. Maybe the feature can be added to v2 if requested enough.

**`dotnet nuget stage push`**

Push a package to staging. The package goes through validation and transitions to Staged on success.

```
dotnet nuget stage push <PACKAGE_PATH> [options]

Arguments:
  PACKAGE_PATH    Path to the .nupkg or .snupkg file to stage.

Options:
  -s, --source <SOURCE>     The package source to use. Defaults to nuget.org.
  --group <GROUP_ID>        Add the package to a staging group. The group must already exist.
  --unlist                   Mark the package as unlisted when promoted.
  --format <FORMAT>         Output format: text (default) or json.
```

Examples:
```
dotnet nuget stage push MyPackage.1.0.0.nupkg
dotnet nuget stage push MyPackage.1.0.0.nupkg --group my-release
dotnet nuget stage push MyPackage.1.0.0.nupkg --group my-release --unlist
```

**`dotnet nuget stage list`**

List staged packages for the authenticated user by group.

```
dotnet nuget stage list [options]

Options:
  -s, --source <SOURCE>     The package source to use. Defaults to nuget.org.
  --group <GROUP_ID>        Filter to packages in a specific group.
  --format <FORMAT>         Output format: text (default) or json.
```

> **`dotnet nuget stage status`**

Reviewer 5: You can consider merging the list and status into 1 command since they basically do the same thing, one for packge, the other for group. 

I think using familiar dotnet CLI verbs would be preferred as well, unless absolutely necessary.

Author: Updated.

Get the status of a specific staged package.

```
dotnet nuget stage status <PACKAGE_ID> <VERSION> [options]

Arguments:
  PACKAGE_ID      The package ID.
  VERSION         The package version.

Options:
  -s, --source <SOURCE>     The package source to use. Defaults to nuget.org.
  --format <FORMAT>         Output format: text (default) or json.
```

**`dotnet nuget stage delete`**

Delete a staged package. Works on packages in any staged state (staged, validating, or failed).

```
dotnet nuget stage delete <PACKAGE_ID> <VERSION> [options]

Arguments:
  PACKAGE_ID      The package ID to delete.
  VERSION         The package version to delete.

Options:
  -s, --source <SOURCE>     The package source to use. Defaults to nuget.org.
  --no-confirm              Skip the confirmation prompt.
  --format <FORMAT>         Output format: text (default) or json.
```

> **`dotnet nuget stage group create`**

Reviewer 5: This is a fun long name :D 

So long it basically has me thinking of alternatives, that are may less wordy. 

I'd honestly consider whether we even need the group creating in the CLI mvp if group promotion requires 2FA. 

Beyond that, I think having a higher level example of what `dotnet nuget stage group` has as sub commands would make it to parse from the spec.

Author: @Nigusu-Allehu

Create a new staging group.

```

> dotnet nuget stage group create <GROUP_ID> [options]

Reviewer 5: I wonder if it's the right time to start "rebranding" the `nuget  push` commands as `package push`. 

cc @baronfel I'd love to get your thoughts on this.

Author: @Nigusu-Allehu

Arguments:
  GROUP_ID        The group identifier. Lowercase letters, numbers, and dashes only.

Options:
  -s, --source <SOURCE>     The package source to use. Defaults to nuget.org.
  --name <DISPLAY_NAME>     Optional display name for the group. Defaults to the group ID.
  --organization <ORG>      Assign ownership to an organization instead of the calling user.
  --format <FORMAT>         Output format: text (default) or json.
```

**`dotnet nuget stage group list`**

> List staging groups for the authenticated user.

Reviewer 6: Is there a group detail option? This would enable automation to track the status of a given stage group.

Author: Updated the spec:
`dotnet nuget stage group list [<GROUP_ID>] [options]`

Options:
  -s, --source <SOURCE>     The package source to use. Defaults to nuget.org.
  --no-confirm              Skip the confirmation prompt.
  --format <FORMAT>         Output format: text (default) or json.
```

**`dotnet nuget stage group add`**

Add a staged package to a group.

```
dotnet nuget stage group add <GROUP_ID> <PACKAGE_ID> <VERSION> [options]

Arguments:
  GROUP_ID        The group to add the package to.
  PACKAGE_ID      The package ID.
  VERSION         The package version.

Options:
  -s, --source <SOURCE>     The package source to use. Defaults to nuget.org.
  --format <FORMAT>         Output format: text (default) or json.
```

**`dotnet nuget stage group remove`**

Remove a staged package from a group (the package remains staged but ungrouped).

```
dotnet nuget stage group remove <GROUP_ID> <PACKAGE_ID> <VERSION> [options]

Arguments:
  GROUP_ID        The group to remove the package from.
  PACKAGE_ID      The package ID.
  VERSION         The package version.

Options:
  -s, --source <SOURCE>     The package source to use. Defaults to nuget.org.
  --format <FORMAT>         Output format: text (default) or json.
```

## Drawbacks

> - **Promotion is not automatable.** Requiring Gallery UI login with step-up 2FA means promotion cannot be scripted or run from CI/CD. This is intentional for security but adds a manual step to release workflows. Someone has to log in and complete a 2FA challenge on release day.

Reviewer 6: I understand the concerns and need for the 2FA challenge. However, since we already have an authenticated human kicking off the release pipeline on the .NET side, this is an added stoppage. It would be great to see if there's a way to achieve the security goal without the additional human touch.

Author: It might be better to view the additional "stoppage" as an additional layer of security. Keep in mind this feature is being developed to help raise the security floor of the entire .NET ecosystem. I sympathize with the fact that the .NET team already has good security practices in place, but many developers and organization do not and I don't think we should pick and choose who has this requirement.
